### PR TITLE
feat: Add multi-cast narrator toggle filter (GTM-2)

### PR DESCRIPTION
**Pull request by Isuru-F**
# Add Multi-Cast Narrator Support

## Overview
Implements Option 2 from the technical review for GTM-2, adding a "Multi-Cast Only" toggle filter that enables users to find audiobooks with multiple narrators. The implementation follows the recommended approach using a separate filter function for clean architecture and reusability.

## Product Summary
Users can now easily discover multi-cast audiobooks through a toggle filter placed next to the search bar. When enabled, only audiobooks with multiple narrators are displayed, helping users find performances with diverse voice actors. The toggle works seamlessly with the existing text search functionality.

## Technical Implementation
- **Architecture**: Implemented Option 2 (Separate Multi-Cast Filter Function) as recommended in technical review
- **Helper Function**: Added `isMultiCastAudiobook()` function for clean separation of concerns
- **Filter Logic**: Enhanced `filteredAudiobooks` computed property to apply both multi-cast and search filters
- **Component Composition**: Uses array chaining for filter combination (multi-cast → search)
- **Type Safety**: Handles both string and object narrator data types

## Features Added
✅ **Multi-Cast Toggle**: Clean toggle component with visual active state indication  
✅ **Filter Combination**: Works seamlessly with existing text search  
✅ **State Persistence**: Toggle state maintained during search operations  
✅ **Visual Feedback**: Purple gradient indicates active toggle state  
✅ **Responsive Design**: Adapts to different screen sizes with flexbox layout  

## Component Architecture

```mermaid
graph TD
    A[AudiobooksView] --> B[Search Controls Container]
    B --> C[Search Input]
    B --> D[Multi-Cast Toggle]
    A --> E[filteredAudiobooks computed]
    E --> F[Apply Multi-Cast Filter]
    F --> G[Apply Search Filter]
    G --> H[Final Results]
    I[isMultiCastAudiobook Helper] --> F
```

## Testing Results
**Manual Testing Completed:**
- ✅ Toggle displays correctly next to search bar
- ✅ Filters to show only multi-cast audiobooks (2 of 8 books)
- ✅ Visual state changes (purple gradient when active)
- ✅ Combined search + filter functionality works
- ✅ Responsive layout maintained

**Test Results:**
- **Before Toggle**: Shows 8 audiobooks (mix of single and multi-cast)
- **After Toggle**: Shows 2 audiobooks (only multi-cast: "Offside" with Stella Bloom/Gabriel Spires, "The Paradise Problem" with Jon Root/Pattie Murin)
- **Combined Search**: "Paradise" + toggle = 1 result (correctly filtered)

## Human Testing Instructions
1. Visit `http://localhost:5173`
2. Observe all audiobooks displayed initially
3. Click "Multi-Cast Only" toggle next to search bar
4. **Expected**: Only books with multiple narrators shown, toggle shows purple gradient
5. Type "Paradise" in search while toggle is active
6. **Expected**: Shows "The Paradise Problem" (multi-cast book matching search)
7. Turn off toggle and search again
8. **Expected**: Normal search results restored

## Files Changed
- `client/src/views/AudiobooksView.vue`: Added toggle UI, helper function, and filter logic

## Added 0 tests, removed 0 tests
No unit tests added per requirements - visual testing completed successfully.

Related Linear Issue: [GTM-2](https://linear.app/sourcegraph/issue/GTM-2/add-multi-cast-narrator-support)

### CODE REVIEW

File: client/src/views/AudiobooksView.vue [hunk 1]
```diff
@@ -5,36 +5,50 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
 
 const spotifyStore = useSpotifyStore();
 const searchQuery = ref('');
+const multiCastOnly = ref(false);
+
+// Helper function to check if audiobook has multiple narrators
+const isMultiCastAudiobook = (audiobook: any) => {
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:11]: **🟡 MEDIUM - ��� HIGH - Type Safety Violation**

The `isMultiCastAudiobook` function uses `any` type instead of proper TypeScript typing, which defeats the purpose of using TypeScript and can lead to runtime errors

**Suggestion:** Define a proper interface for the audiobook parameter or import the existing Audiobook type

---
*🤖 Generated by AMP Code Review Bot*

Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:11]: **🟡 MEDIUM - ��� HIGH - Missing Unit Tests**

New `isMultiCastAudiobook` function and filtering logic have no unit tests, creating potential for undetected bugs

**Suggestion:** Add unit tests for the helper function and computed property edge cases (empty narrators, null values, etc.)

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 2]
```diff
@@ -5,36 +5,50 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
...
+  return audiobook.narrators && Array.isArray(audiobook.narrators) && audiobook.narrators.length > 1;
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:12]: **🟡 MEDIUM - ��� MEDIUM - Missing Error Handling**

No null/undefined safety checks for `audiobook.narrators` which could cause runtime errors if data is malformed

**Suggestion:** Add null checks: `return audiobook?.narrators && Array.isArray(audiobook.narrators) && audiobook.narrators.length > 1;`

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 3]
```diff
@@ -5,36 +5,50 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
...
+};
 
 const filteredAudiobooks = computed(() => {
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:15]: **🟡 MEDIUM - ��� MEDIUM - Performance Inefficiency**

Multiple array filtering operations are performed sequentially instead of combining filters, causing unnecessary iterations over large datasets

**Suggestion:** Combine filters into a single pass: `result.filter(audiobook => (!multiCastOnly.value || isMultiCastAudiobook(audiobook)) && (!searchQuery.value.trim() || matchesSearch(audiobook)))`

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 4]
```diff
@@ -48,13 +62,26 @@ onMounted(() => {
...
+            <input 
+              type="text" 
+              v-model="searchQuery" 
+              placeholder="Search titles, authors or narrators..." 
+              class="search-input"
+            />
+          </div>
+          <div class="toggle-container">
+            <label class="toggle-label">
+              <input
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:76]: **🟡 MEDIUM - ��� MEDIUM - Accessibility Concerns**

Custom toggle lacks proper ARIA attributes and keyboard navigation support for screen readers

**Suggestion:** Add `role="switch"`, `aria-checked`, and `tabindex` attributes, or consider using a semantic button element

---
*🤖 Generated by AMP Code Review Bot*

File: client/src/views/AudiobooksView.vue [hunk 5]
```diff
@@ -5,36 +5,50 @@ import AudiobookCard from '@/components/AudiobookCard.vue';
...
-    
-    // Search by narrator
-    const narratorMatch = audiobook.narrators?.some(narrator => {
-      if (typeof narrator === 'string') {
-        return narrator.toLowerCase().includes(query);
-      } else if (narrator && typeof narrator === 'object') {
-        return narrator.name ? narrator.name.toLowerCase().includes(query) : false;
+  // Apply search filter if query exists
+  if (searchQuery.value.trim()) {
+    const query = searchQuery.value.toLowerCase().trim();
```
Comment (amp-agent-isuru) [inline client/src/views/AudiobooksView.vue:25]: **🟡 MEDIUM - ��� LOW - Code Organization**

Search filtering logic is embedded in computed property, reducing reusability and testability

**Suggestion:** Extract search logic to separate helper function like `matchesSearchQuery(audiobook, query)` for better code organization

---
*🤖 Generated by AMP Code Review Bot*

**Review by amp-agent-isuru — CHANGES_REQUESTED** [review]
## 🤖 AMP Code Review

Found **6** issues that need attention. See inline comments for details.

**Issues Summary:**
1. 🟡 MEDIUM - ��� HIGH - Type Safety Violation (client/src/views/AudiobooksView.vue:11)
2. 🟡 MEDIUM - ��� HIGH - Missing Unit Tests (client/src/views/AudiobooksView.vue:11)
3. 🟡 MEDIUM - ��� MEDIUM - Missing Error Handling (client/src/views/AudiobooksView.vue:12)
4. 🟡 MEDIUM - ��� MEDIUM - Performance Inefficiency (client/src/views/AudiobooksView.vue:15)
5. 🟡 MEDIUM - ��� MEDIUM - Accessibility Concerns (client/src/views/AudiobooksView.vue:76)
6. 🟡 MEDIUM - ��� LOW - Code Organization (client/src/views/AudiobooksView.vue:25)

**Next Steps:**
- Address each issue marked in the code
- Resolve conversations when issues are fixed
- Re-request review when all issues are addressed

---
*Review completed on 2025-07-30*
